sway/desktop/output.c

```diff
@@ -522,17 +540,40 @@ static void damage_handle_frame(struct wl_listener *listener, void *data) {
 	pixman_region32_init(&damage);
 	if (!wlr_output_damage_attach_render(output->damage,
 			&needs_frame, &damage)) {
-		return;
+		return 0;
 	}
 
 	if (needs_frame) {
+		struct timespec now;
+		clock_gettime(CLOCK_MONOTONIC, &now);
+
 		output_render(output, &now, &damage);
 	}
 
 	pixman_region32_fini(&damage);
 
-frame_done:
+	return 0;
+}
+
+static void damage_handle_frame(struct wl_listener *listener, void *data) {
+	struct sway_output *output =
+		wl_container_of(listener, output, damage_frame);
+	if (!output->enabled || !output->wlr_output->enabled) {
+		return;
+	}
+
+	int delay = output->refresh_nsec / 1000000 - output->max_render_time;
+
+	if (output->max_render_time == 0 || delay < 1) {
+		output_repaint_timer_handler(output);
+	} else {
+		output->wlr_output->block_idle_frame = true;
+		wl_event_source_timer_update(output->repaint_timer, delay);
```

Comment:
So, the convenient thing about doing it here is that it's a single spot where the frames are either drawn normally (with the delay disabled) or scheduled (with the delay enabled). It is also where surfaces are sent the frame callback.

If the output frames are scheduled from the present event handler, then:
- Assuming that the present event handler, if it's called, is called right before the frame handler, the logic for present would be "schedule a frame" and the logic for frame would be "if a frame isn't scheduled, render a frame right away". However, if it's possible for the present event handler to be called later, I'm not sure how to handle the interaction, since we always want to render from the frame handler, and we wouldn't know there if the present event handler will be called shortly after.
- There are cases where the present event handler is called, but the frame event handler isn't, for example, I'm assuming, when the output is disabled (the last frame is presented, but no new frame is requested). The present event handler doesn't know about this, so how should it know whether to schedule a frame?
- The surface frame callback sending would also then be duplicated between the frame handler (if surface delay is disabled) and the present handler (if the delay is enabled, since then it should also be computed from the presentation time), which raises all the same issues.

Reply:
Sorry about the delay. I'll prioritize this pull request from now on so that we can get this merged.

>It is also where surfaces are sent the frame callback.

Or where the frame callback is scheduled, I assume.

>Assuming that the present event handler, if it's called, is called right before the frame handler, the logic for present would be "schedule a frame" and the logic for frame would be "if a frame isn't scheduled, render a frame right away". However, if it's possible for the present event handler to be called later, I'm not sure how to handle the interaction, since we always want to render from the frame handler, and we wouldn't know there if the present event handler will be called shortly after.

The present event is fired on presentation, so always before the frame callback.

>There are cases where the present event handler is called, but the frame event handler isn't, for example, I'm assuming, when the output is disabled (the last frame is presented, but no new frame is requested). The present event handler doesn't know about this, so how should it know whether to schedule a frame?

I don't think it's a big deal if we schedule a frame and rendering ends up being cancelled because the output is disabled. We already check whether the output is disabled in the rendering function.

>The surface frame callback sending would also then be duplicated between the frame handler (if surface delay is disabled) and the present handler (if the delay is enabled, since then it should also be computed from the presentation time), which raises all the same issues.

Isn't this just a matter of calling `send_frame_done` from these two places, with a `if`?

Reply:
> The present event is fired on presentation, so always before the frame callback.

So, then, a simple solution is possible of storing the presentation info in the presentation callback, and using that info to compute the delays in the frame callback, otherwise leaving the code as is, right?

Reply:
All right, let's do this. It'll be easy enough to change if needed anyway.